pkg/wallet/utils.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package wallet
+
+import (
+	"crypto"
+	"encoding/binary"
+	"errors"
+
+	josecipher "github.com/square/go-jose/v3/cipher"
+	chacha "golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/curve25519"
+)
+
+// errEmptyRecipients is used when recipients list is empty
+var errEmptyRecipients = errors.New("empty recipients")
+
+// errInvalidKeypair is used when a keypair is invalid
+var errInvalidKeypair = errors.New("invalid keypair")
+
+// VerifyKeys is a utility function that verifies if sender key pair and recipients keys are valid (not empty)
+func VerifyKeys(sender KeyPair, recipients [][]byte) error {
+	if len(recipients) == 0 {
+		return errEmptyRecipients
+	}
+
+	if !IsKeyPairValid(sender) {
+		return errInvalidKeypair
+	}
+
+	if !IsChachaKeyValid(sender.Priv) || !IsChachaKeyValid(sender.Pub) {
+		return ErrInvalidKey
+	}
+	return nil
+}
+
+// IsChachaKeyValid will return true if key size is the same as chacha20poly1305.keySize
+// false otherwise
+func IsChachaKeyValid(key []byte) bool {
+	return len(key) == chacha.KeySize
+}
+
+// KeyPair represents a private/public key pair
+type KeyPair struct {
+	// Priv is a private key
+	Priv []byte
+	// Pub is a public key
+	Pub []byte
+}
+
+// IsKeyPairValid is a utility function that validates a KeyPair
+func IsKeyPairValid(kp KeyPair) bool {
+	if kp.Priv == nil || kp.Pub == nil {
+		return false
+	}
+
+	return true
+}
+
+// Derive25519KEK is a utility function that will derive an ephemeral symmetric key (kek) using fromPrivKey and toPubKey
+func Derive25519KEK(alg, apu []byte, fromPrivKey, toPubKey *[chacha.KeySize]byte) ([]byte, error) { // nolint:lll
```

Comment:
Is this used outside the package?

Reply:
yes

Reply:
It's used in the Crypter - DeriveKEK was split into a util function because the crypter needs to be able to pass an ephemeral private key in at one point.

Reply:
It still seems odd that the **generic** wallet package has a special ed25519 method.

Reply:
I will move utils.go to pkg/internal/cryptoutil to separate the utility logic from the wallet.. in the end, they are internal to our go framework and should not be exposed in the wallet.

Reply:
done... moved utils.go to `pkg/internal/cryptoutil` package and updated all its references.